MCO-1877: MCO-1879: MCO-1882: MCO-1884: Implement boot image skew enforcement MVP

djoshy opened the pull request:
This PR integrates the boot image skew enforcement API introduced in https://github.com/openshift/api/pull/2357.  This involves the following changes:
- The operator now populates the `bootImageSkewEnforcementStatus` field in the `MachineConfiguration` object based on `spec.bootImageSkewEnforcement`, platform defaults and cluster version. 
- The boot image controller will now update the current boot image value in `bootImageSkewEnforcementStatus` on a successful boot image update. Note that this requires the skew enforcement to be set to `Automatic` mode, and all machinesets to be opt-ed in for boot image updates.
- The operator sets Upgradeable=False when it detects the cluster is out of skew, determined by comparing the boot image values in bootImageSkewEnforcementStatus against the MCO's hardcoded skew limits. Before performing this check, the operator first verifies that the controller is neither in an error state nor currently performing boot image updates. If the controller is in an error state, the operator sets Upgradeable=False and propagates that error instead of proceeding with the skew check. If the controller is mid-update, the operator defers the skew check until later; this is to avoid race conditions.
- Some unit tests have been added to `sync_test.go` and `status_test.go` to verify the above mechanisms.

### Verifying API behavior
This verification will have to be done based on the platform. If the platform:

- **supports boot image updates and it is on by default(AWS and GCP at the time of writing)**, i.e. `status.managedBootImagesStatus` is set to `All` if `spec.managedBootImages` is empty. Then, skew enforcement status will be set to `Automatic`, with a boot image version estimated from cluster version. Then, the boot image controller will perform a sync which will update the boot image(if required) and after all resources have been successfully updated, it will update the boot image value stored in the skew enforcement status. The value set will be the OCP `releaseVersion` described by the `coreos-bootimages` configmap. Here's an example:
```
  spec:
    logLevel: Normal
    managementState: Managed
    operatorLogLevel: Normal
  status:
    bootImageSkewEnforcementStatus:
      automatic:
        ocpVersion: 4.21.0
      mode: Automatic
    conditions:
    - lastTransitionTime: "2025-11-19T22:06:06Z"
      message: Reconciled 3 of 3 MAPI MachineSets | Reconciled 0 of 0 ControlPlaneMachineSets
        | Reconciled 0 of 0 CAPI MachineSets | Reconciled 0 of 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateProgressing
    - lastTransitionTime: "2025-11-19T22:06:07Z"
      message: 0 Degraded MAPI MachineSets | 0 Degraded ControlPlaneMachineSets |
        0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateDegraded
    managedBootImagesStatus:
      machineManagers:
      - apiGroup: machine.openshift.io
        resource: machinesets
        selection:
          mode: All
```

- **supports boot image updates, but is not on by default(vsphere and Azure at the time of writing)** i.e. `status.managedBootImagesStatus` is set to `None` if `spec.managedBootImages` is empty. Then, skew enforcement status will be set to `Manual`, with a boot image version estimated from cluster version. The object would now look like this:
```
  spec:
    logLevel: Normal
    managementState: Managed
    operatorLogLevel: Normal
  status:
    bootImageSkewEnforcementStatus:
      manual:
        mode: OCPVersion
        ocpVersion: 4.21.0
      mode: Manual
    conditions:
    - lastTransitionTime: "2025-11-19T22:06:06Z"
      message: Reconciled 0 of 0 MAPI MachineSets | Reconciled 0 of 0 ControlPlaneMachineSets
        | Reconciled 0 of 0 CAPI MachineSets | Reconciled 0 of 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateProgressing
    - lastTransitionTime: "2025-11-19T22:06:07Z"
      message: 0 Degraded MAPI MachineSets | 0 Degraded ControlPlaneMachineSets |
        0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateDegraded
    managedBootImagesStatus:
      machineManagers:
      - apiGroup: machine.openshift.io
        resource: machinesets
        selection:
          mode: None
```
The admin can choose to opt-in for boot image updates in this case(set `spec.ManagedBootImages` to `All`), and the operator should automatically switch the skew enforcement status to `Automatic`, with the appropriate boot image version. This would mean the object would finally look like this:
```
  spec:
    logLevel: Normal
    managementState: Managed
    operatorLogLevel: Normal
    managedBootImages:
      machineManagers:
      - apiGroup: machine.openshift.io
        resource: machinesets
        selection:
          mode: All
  status:
    bootImageSkewEnforcementStatus:
      automatic:
        ocpVersion: 4.21.0
      mode: Automatic
    conditions:
    - lastTransitionTime: "2025-11-19T22:06:06Z"
      message: Reconciled 3 of 3 MAPI MachineSets | Reconciled 0 of 0 ControlPlaneMachineSets
        | Reconciled 0 of 0 CAPI MachineSets | Reconciled 0 of 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateProgressing
    - lastTransitionTime: "2025-11-19T22:06:07Z"
      message: 0 Degraded MAPI MachineSets | 0 Degraded ControlPlaneMachineSets |
        0 Degraded CAPI MachineSets | 0 CAPI MachineDeployments
      reason: BootImageConfigMapAdded
      status: "False"
      type: BootImageUpdateDegraded
    managedBootImagesStatus:
      machineManagers:
      - apiGroup: machine.openshift.io
        resource: machinesets
        selection:
          mode: All
```
- **does not support boot image updates(all other platforms at the time of writing)**  i.e. `status.managedBootImagesStatus` is empty and `spec.managedBootImages` cannot be set by the admin. Then, skew enforcement status will be set to `Manual`, with a boot image version estimated from cluster version. The object would now look like this:
```
  spec:
    logLevel: Normal
    managementState: Managed
    operatorLogLevel: Normal
  status:
    bootImageSkewEnforcementStatus:
      manual:
        mode: OCPVersion
        ocpVersion: 4.21.0
      mode: Manual
```
In this case, the admin is expected to manually perform boot image updates and then add a spec field like so:
```
spec:
  bootImageSkewEnforcement:
    mode: Manual
    manual:
      mode: OCPVersion
      ocpVersion: 4.21.2
```
The operator should then update the status to include this:
```
spec:
  bootImageSkewEnforcement:
    mode: Manual
    manual:
      mode: OCPVersion
      ocpVersion: 4.21.2
status:
  bootImageSkewEnforcementStatus:
      mode: OCPVersion
      ocpVersion: 4.21.2
```
The above snippet is if an admin had chosen to record the `OCPVersion`. In manual mode, the admin can also choose to to store the `RHCOSVersion`, like so:
```
spec:
  bootImageSkewEnforcement:
    mode: Manual
    manual:
      mode: RHCOSVersion
      rhcosVersion: 9.0.20251023-0
status:
  bootImageSkewEnforcementStatus:
    mode: Manual
    manual:
      mode: RHCOSVersion
      rhcosVersion: 9.0.20251023-0
```
Note that only one of RHCOSVersion or OCPVersion is permitted in `Manual` mode. 

The admin can also choose to disable skew enforcement altogether by setting it `None` mode in spec.
```
spec:
  bootImageSkewEnforcement:
    mode: None
status:
  bootImageSkewEnforcementStatus:
    mode: None
```

### Verifying upgrade block

Upgrades will be blocked when the cluster is to determined out of skew. This mechanism works the same way in manual and automatic mode, although it is likely easier to verify in manual mode. The current thresholds for a skew violation is set to when OCP first moved to RHEL9, which corresponds to RHEL version 9.2 and OCP version 4.13.0. The operator will perform semver comparisons of these thresholds against the boot image versions stored in `bootImageSkewEnforcementStatus` and set `Upgradeable=False` if necessary. To verify this, first set the mode to Manual with an out of skew boot image version like so:
```
  spec:
    bootImageSkewEnforcement:
      manual:
	mode: RHCOSVersion
        rhcosVersion: 9.0.20251023-0
      mode: Manual
```
Now, examine the `machine-config` CO object's conditions field, it should indicate an issue preventing upgrades like so:
```
$ oc get co machine-config -o yaml
...
  - lastTransitionTime: "2025-11-20T15:15:12Z"
    message: 'Upgrades have been disabled because the cluster is using RHCOS boot
      image version 9.0.20251023-0(RHEL version: 9.0), which is below the minimum
      required RHEL version 9.2. To enable upgrades, please update your boot images
      following the documentation at [TODO: insert link], or disable boot image skew
      enforcement at [TODO: insert link]'
    reason: ClusterBootImageSkewError
    status: "False"
    type: Upgradeable
```
Next, set the boot image to one within the skew limits:
```
  spec:
    bootImageSkewEnforcement:
      manual:
	mode: RHCOSVersion
        rhcosVersion: 9.2.20251023-0
      mode: Manual
```
Then, the `Upgradeable`  condition should be restored back to `True`
```
  - lastTransitionTime: "2025-11-20T15:19:25Z"
    reason: AsExpected
    status: "True"
    type: Upgradeable
```

These set of steps can be repeated with the OCPVersion specified too. This comparison should only take place in `Automatic` and `Manual` mode however, as `Automatic` is only permitted on the status side, I don't think there is an easy way to test that(other than the units I've included). 

In `None` mode, this version check should not take place.

Some caveats to note about `Automatic` mode: 

1. The admin is not permitted to use `Automatic` mode within the spec. This is in an intentional choice because only the MCO will always be able to self determine if a platform is eligible for automatic skew enforcement. 
2. In `Automatic` mode, API validations will prevent changing the boot image configuration to a setting other than `All`. To change the boot image configuration, the admin is first expected to go to `Manual` skew enforcement mode and then attempt to change the boot image configuration of the cluster.
3. In `Automatic` mode, if any machinesets are skipped for boot image updates(for example a marketplace or an unknown boot image was detected in any of the machinesets), the boot image controller will not update the boot image value stored in bootImageEnforcementStatus. This is because the cluster cannot be considered up to date on boot image if even one of the machine resources are out of skew. 
4. In `Automatic` mode, the operator will only populate the OCPVersion. This is because each platform may not have the same RHCOS version of the boot image(for example, across marketplace streams) in a given release, and it would involve a lot of per-platform piping to correctly track the RHCOS version per machineset within the boot image controller. I did not deem this to be worth the effort, but am open to implementing that later if the need arises.

djoshy committed:
bootimage: update skew enforcement status

This commit updates the machine-set-boot-image controller to track and
update the BootImageSkewEnforcementStatus when in Automatic mode.

djoshy committed:
operator: populate skew enforcement status

This change adds logic to populate the BootImageSkewEnforcementStatus
field in the MachineConfiguration status based on spec configuration,
platform support, and cluster version information.

djoshy committed:
operator: add skew enforcement status unit tests

This adds new unit tests for TestSyncMachineConfiguration to test the
BootImageSkewEnforcementStatus sync logic added in the previous commit.

djoshy committed:
operator: block upgrades on skew violation

This commit implements upgrade blocking when boot image version skew
exceeds acceptable limits, via the ClusterOperator Upgradeable
condition.

djoshy committed:
operator: add upgrade block unit tests

This commit adds unit tests for the new Upgradeable guards added in the
previous commit.

djoshy commented:
/retest-required

isabella-janssen commented on pkg/controller/common/constants.go:
```diff
@@ -152,6 +152,14 @@ const (
+	// Current Boot Image Skew Limits
+	// Note: Update units in status_test.go when the following are bumped
+	RHCOSVersionBootImageSkewLimit = "9.2"
+	OCPVersionBootImageSkewLimit   = "4.13.0"
```
How will we remember to bump these?

isabella-janssen commented on pkg/operator/status.go:
```diff
@@ -631,3 +645,170 @@ func machineConfigPoolStatus(fgHandler ctrlcommon.FeatureGatesHandler, pool *mcf
+	case opv1.BootImageSkewEnforcementModeStatusNone:
+		// TODO: Set a low level prom alert to set scaling risk
+		klog.V(4).Infof("evaluating boot image skew enforcement: mode set to None")
+		return false, "", nil
```
Nit: Attach to a Jira if possible.

isabella-janssen commented on pkg/operator/status.go:
```diff
@@ -631,3 +645,170 @@ func machineConfigPoolStatus(fgHandler ctrlcommon.FeatureGatesHandler, pool *mcf
+
+	if skewLimitExceeded {
+		return true, fmt.Sprintf("Upgrades have been disabled because %s. To enable upgrades, please update your boot images following the documentation at [TODO: insert link], or disable boot image skew enforcement at [TODO: insert link]", skewLimitExceededMessage), nil
+	}
```
Nit: Attach to a Jira if possible.

isabella-janssen reviewed (commented):
Generally this looks good to me and makes sense. There are a couple TODO sections--are those to be completed in this PR or some follow up work?

djoshy committed:
operator: verify boot image controller state

This commit ensures that the boot image controller state is acceptable before checking the skew. This check is only done in Automatic mode.

yuqi-zhang commented on pkg/controller/bootimage/boot_image_controller.go:
```diff
@@ -446,8 +448,10 @@ func (ctrl *Controller) updateMachineConfiguration(oldMC, newMC interface{}) {
-	// Only take action if the there is an actual change in the MachineConfiguration's ManagedBootImagesStatus
-	if reflect.DeepEqual(oldMachineConfiguration.Status.ManagedBootImagesStatus, newMachineConfiguration.Status.ManagedBootImagesStatus) {
+	// Only take action if the there is an actual change in the MachineConfiguration's ManagedBootImagesStatus or BootImageSkewEnforcementStatus(when bootimageskewenforcement feature gate is enabled)
+	if reflect.DeepEqual(oldMachineConfiguration.Status.ManagedBootImagesStatus, newMachineConfiguration.Status.ManagedBootImagesStatus) &&
```
This... reads slightly odd. Based on the comment I think it might be clearer if it evaluated ( change in ManagedBootImagesStatus) || ((FeatureGateBootImageSkewEnforcement && change in BootImageSkewEnforcementStatus)) ? (if I understood that correctly)

I think what you have also is correct since && takes precedent plus the `!` on FeatureGateBootImageSkewEnforcement, but was a bit logically confusing to read

yuqi-zhang commented on pkg/controller/bootimage/boot_image_controller.go:
```diff
@@ -523,33 +527,77 @@ func (ctrl *Controller) updateConditions(newReason string, syncError error, targ
-			return nil
-		}); err != nil {
-			klog.Errorf("error updating MachineConfiguration status: %v", err)
+	klog.V(2).Infof("MachineConfiguration status update: %v", mcopStatus)
```
Did you mean to set this at a lower verbosity? Our default is 2 anyways. (looking at a job run of this PR: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/5428/pull-ci-openshift-machine-config-operator-main-e2e-gcp-op-1of2/2008265275299860480/artifacts/e2e-gcp-op-1of2/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-controller-6d88bcc7f7-79x74_machine-config-controller.log , this would get logged 36 times, which I guess isn't that bad?)

yuqi-zhang commented on pkg/operator/status.go:
```diff
@@ -631,3 +645,206 @@ func machineConfigPoolStatus(fgHandler ctrlcommon.FeatureGatesHandler, pool *mcf
+	skewLimitExceeded := false
+	skewLimitExceededMessage := ""
+
+	switch mcop.Status.BootImageSkewEnforcementStatus.Mode {
```
Could you help me with my understanding of a scenario: if I have a cluster with marketplace images, the enforcement  status on MachineConfiguration object would never get updated since the machineset is skipped, and any skipped machineset causes the enforcement status to not get updated via the `updateClusterBootImage()` call, which then means that the status deferences below would cause... a panic? Or is there something else preventing that?

I guess the more overarching question is what we're doing with platforms/scenarios not covered by our updates, do we plan to have all documentation ready for for all scenarios when this goes GA? Or is there a softer approach we can take based on platform? (forgive me if this is already discussed/covered, I don't quite recall the details from pre-holidays)

yuqi-zhang replied:
Thanks for the summary, lgtm based on our discussions

yuqi-zhang commented on pkg/operator/sync.go:
```diff
@@ -2461,3 +2466,75 @@ func (optr *Operator) syncPreBuiltImageMachineConfigs() error {
+				Manual: mcop.Spec.BootImageSkewEnforcement.Manual,
+			}
+		} else { // only other possible opinion is "None"
+			newMachineConfigurationStatus.BootImageSkewEnforcementStatus = apihelpers.GetSkewEnforcementStatusNone()
```
(nit) I guess for future proofing we can probably just set this to the spec value instead of hard-coding to None, but realistically shouldn't make a difference

djoshy replied:
Good point, the annoying bit here is that spec <-> status for skew enforcement is not a direct copy, so we'd be transposing the value in some fashion. Unless you meant that the fact that we're using an if/else ladder is confusing and we should use a switch case?

yuqi-zhang replied:
Right, sorry, I forget that they're technically different types with the same values. Let's keep it as is for now.

yuqi-zhang commented on pkg/operator/sync.go:
```diff
@@ -2461,3 +2466,75 @@ func (optr *Operator) syncPreBuiltImageMachineConfigs() error {
+			newMachineConfigurationStatus.BootImageSkewEnforcementStatus = apihelpers.GetSkewEnforcementStatusNone()
+		}
+	} else if supportsBootImageUpdates {
+		// If an "All" option is specified and BootImageSkewEnforcementStatus is empty or not set to Automatic => set Mode to Automatic.
```
Does this not conflict with updateClusterBootImage() at all? I thought that was also updating the MachineConfiguration status

yuqi-zhang reviewed (commented):
The general structure makes sense to me, some questions inline

djoshy commented on pkg/controller/bootimage/ms_helpers.go:
```diff
@@ -126,15 +136,15 @@ func (ctrl *Controller) syncMAPIMachineSet(machineSet *machinev1beta1.MachineSet
 		// A later sync loop will pick it up once the annotation is added
 		if strings.Contains(err.Error(), "no architecture annotation found") {
-			return nil
+			return true, nil
```
Right, this is an install time thing, so the annotation should settle once that is complete. 

In terms of skew limits, it's not possible for the cluster to be out of skew due to this particular return, since the operator initializes the skew status with the install time OCP version(which should be within skew limits during install and a few versions into the future, presumably)

djoshy commented on pkg/controller/bootimage/ms_helpers.go:
```diff
@@ -76,23 +77,32 @@ func (ctrl *Controller) syncMAPIMachineSets(reason string) {
 			ctrl.mapiStats.erroredCount++
 		}
+		if patchSkipped {
+			ctrl.mapiStats.skippedCount++
```
Ah, good catch - I should definitely be resetting this along with the others at line 70. 

As for the type, I think I followed the existing pattern here, and maybe thought about potentially adding this field to the conditions message(https://github.com/openshift/machine-config-operator/blob/36dcbbc1ae89b6f0a9de54ad486f7120f32452a8/pkg/controller/bootimage/boot_image_controller.go#L562), but never got around to it. What do you think? Enhancing the conditions message might help while debugging for those marketplace cases...

yuqi-zhang replied:
I'm fine with the typing if it helps add to the status message :+1:

yuqi-zhang commented:
Thanks for the updates. Structure lgtm, will let others take a look/test as well.

sergiordlr commented:
Verified using IPI on AWS, GCP, Azure and Vsphere

Automatic skew was configured by MCO in AWS and GCP and Manual skew was configured by MCO in Azure and Vsphere.

We tested that the right version was used by the skew process by scaling down the CVO and manually editing the history in the clusterversion resource. MCO is correctly reporting the oldest version in clusterversion.status.history in the skew version, and it is correctly updating the value to the latest version when the bootimage cycle is successfully executed.

In https://github.com/openshift/machine-config-operator/pull/5547  we can see the automation for the tests that were executed to verify this PR (apart from manually hacking the history in clusterversion).

There is a pending test: upgrading a 4.12 cluster up to 4.22. We are working on it, nevertheless it will take some time and should not block this PR. If any problem is found in this test it can be reported as an issue after merging the code.

/verified by @sergiordlr 


PS: to manually hack the clusterversion history use: `oc edit clusterversion version --subresource=status`

djoshy commented:
/hold

/payload 4.22 nightly blocking

yuqi-zhang reviewed (approved):
/lgtm

I think we've covered most of the main concerns, and we can iterate on some details (e.g. skew limits) as followups since this is still behind TP

djoshy commented:
/test all
